packages/core/src/lib/stream_manager/stream_manager.ts

```diff
@@ -1,47 +1,46 @@
-import type { PeerUpdate, Stream } from "@libp2p/interface";
-import type { Peer, PeerId } from "@libp2p/interface";
-import { Libp2p } from "@waku/interfaces";
+import type {
+  Connection,
+  Peer,
+  PeerId,
+  PeerUpdate,
+  Stream
+} from "@libp2p/interface";
+import type { Libp2p } from "@waku/interfaces";
 import { Logger } from "@waku/utils";
 
 import { selectConnection } from "./utils.js";
 
 const CONNECTION_TIMEOUT = 5_000;
-const RETRY_BACKOFF_BASE = 1_000;
-const MAX_RETRIES = 3;
 
 export class StreamManager {
-  private readonly streamPool: Map<string, Promise<Stream | void>>;
   private readonly log: Logger;
 
+  private ongoingCreation: Set<string> = new Set();
+  private streamPool: Map<string, Promise<void>> = new Map();
+
   public constructor(
     public multicodec: string,
     public getConnections: Libp2p["getConnections"],
     public addEventListener: Libp2p["addEventListener"]
   ) {
     this.log = new Logger(`stream-manager:${multicodec}`);
-    this.streamPool = new Map();
     this.addEventListener("peer:update", this.handlePeerUpdateStreamPool);
   }
 
   public async getStream(peer: Peer): Promise<Stream> {
-    const peerIdStr = peer.id.toString();
-    const streamPromise = this.streamPool.get(peerIdStr);
+    const peerId = peer.id.toString();
 
-    if (!streamPromise) {
-      return this.createStream(peer);
-    }
+    const scheduledStream = this.streamPool.get(peerId);
+    this.streamPool.delete(peerId);
+    await scheduledStream;
 
-    this.streamPool.delete(peerIdStr);
-    this.prepareStream(peer);
+    const stream = this.getStreamForCodec(peer.id);
```

Comment:
if `stream = await connection.newStream(this.multicodec);` in `createStream` succeeds - then it will be attached to `connection` and hence `this.getStreamForCodec` will return it.

This is a bit convoluted at first but:
- we need to keep our single source of truth - `connection` object;
- we need to have a lock so that multiple events schedule only one redundant stream creation;